File: 2021/archived/Day14/Day14.fs
module Day14

open System.IO

let insert seq instructions steps =
    let count state _ =
        state
        |> Seq.map (fun (freq as (pair, count)) ->
            let chars = pair |> Seq.toArray |> Array.map string

            match instructions |> Map.tryFind pair with
            | Some (v) ->
                [| (chars.[0] + v, count)
                   (v + chars.[1], count) |]
            | _ -> [| freq |])

        |> Seq.concat
        |> Seq.groupBy fst
        |> Seq.map (fun (pair, counts) -> (pair, counts |> Seq.sumBy snd))

    [| 1 .. steps |] |> Array.fold count seq

let solve text instructions steps =
    let last = text |> Seq.last

    let freqs =
        text
        |> Seq.windowed 2
        |> Seq.map (Seq.map string >> String.concat "")
        |> Seq.countBy id
        |> Seq.map (fun (pair, count) -> (pair, int64 count))


    let result =
        insert freqs instructions steps
        |> Seq.groupBy (fun (pair, _) -> pair |> Seq.toArray |> Array.head)
        |> Seq.map (fun (c, counts) ->
            let s = counts |> Seq.sumBy snd
            if c = last then s + 1L else s)

    let max = result |> Seq.max
    let min = result |> Seq.min
    max - min

let text = "CKKOHNSBPCPCHVNKHFFK"

let instructions =
    File.ReadAllLines "instructions.txt"
    |> Array.map (fun s ->
        let pair = s.Split(" -> ")
        (pair.[0], pair.[1]))
    |> Map.ofArray

solve text instructions 10
|> printfn "First part: %d"

solve text instructions 40
|> printfn "First part: %d"

File: 2021/archived/Day14/Day14.scala
import scala.io.Source

@main def day14 =
    def insert(seq: Map[String, Long], instructions: Map[String, String])(
        steps: Int
    ) =
        (1 to steps).foldLeft(seq)((state, _) =>
            state
                .map(freq =>
                    val (pair, count) = freq
                    val Array(l, r) = pair.toCharArray
                    instructions.get(pair) match
                        case Some(value) =>
                            Array((l + value, count), (value + r, count))
                        case _ => Array(freq)
                )
                .flatMap(identity)
                .groupBy(_._1)
                .mapValues(pair => pair.map(_._2).sum)
                .toMap
        )

    def solve(seq: String, instructions: Map[String, String])(steps: Int) =
        val last = seq.last
        val init = seq
            .sliding(2)
            .map(x => (x, 1L))
            .toVector
            .groupBy(_._1)
            .mapValues(pair => pair.map(_._2).sum).toMap

        val result = (insert(init, instructions)(steps))
            .groupBy(freq =>
                val (pair, count) = freq
                val Array(l, r) = pair.toCharArray
                l
            )
            .map((c, m) =>
                val s = m.map(_._2).sum
                if c == last then (c, s + 1) else (c, s)
            )
        val max = result.maxBy(_._2)._2
        val min = result.minBy(_._2)._2
        max - min

    val instructions = Source
        .fromFile("instructions.txt")
        .getLines
        .map(line =>
            val Array(l, r) = line.split(" -> ")
            (l, r)
        )
        .toMap

    val text = "CKKOHNSBPCPCHVNKHFFK"

    println(s"First part: ${solve(text, instructions)(10)}")
    println(s"Second part: ${solve(text, instructions)(40)}")
